org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/internal/CommentWithSpacesFormatterTest.xtend

```diff
@@ -0,0 +1,191 @@
+/*******************************************************************************
+ * Copyright (c) 2017 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.formatting2.internal
+
+import com.google.inject.Inject
+import org.eclipse.xtext.formatting2.IFormattableDocument
+import org.eclipse.xtext.formatting2.IHiddenRegionFormatting
+import org.eclipse.xtext.formatting2.ITextReplacer
+import org.eclipse.xtext.formatting2.ITextReplacerContext
+import org.eclipse.xtext.formatting2.internal.formattertestlanguage.IDList
+import org.eclipse.xtext.formatting2.internal.tests.FormatterTestLanguageInjectorProvider
+import org.eclipse.xtext.formatting2.regionaccess.ITextRegionExtensions
+import org.eclipse.xtext.formatting2.regionaccess.ITextSegment
+import org.eclipse.xtext.testing.InjectWith
+import org.eclipse.xtext.testing.XtextRunner
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * @author Moritz Eysholdt - Initial contribution and API
+ */
+@RunWith(XtextRunner)
+@InjectWith(FormatterTestLanguageInjectorProvider)
+class CommentWithSpacesFormatterTest {
+
+	@Inject extension GenericFormatterTester
+
+	static class TabsAndSpacesSupportingFormatter extends GenericFormatter<IDList> {
+		override format(IDList model, ITextRegionExtensions regionAccess, extension IFormattableDocument document) {
+			model.regionFor.keyword("idlist").append[oneSpace].append[newLines = 0 highPriority]
+		}
+
+		override ITextReplacer createWhitespaceReplacer(ITextSegment hiddens, IHiddenRegionFormatting formatting) {
+			return new TabAndSpacesSupportingWhiteSpaceReplacer(hiddens, formatting);
+		}
+	}
+
+	@Test def void SL_inline() {
+		assertFormatted[
+			toBeFormatted = '''
+				idlist  //x
+				a
+			'''
+			formatter = new TabsAndSpacesSupportingFormatter()
+			// space before a because we replace by newline + indentation + space
+			// in the TabAndSpacesSupportingWhiteSpaceReplacer
```

Comment:
where does the `indentation` come from?
I cannot seem to find it in the `format(IDList..)` impl above.
It looks like the tokenization should be 
```
<idList><space><sl-comment incl newline since it's part of the token>
<a>
```

Reply:
This comes from the `append[oneSpace]`.

The document looks like:

```
idlist<6|  //x
>a
--------------------------------------------------------------------------------
6 6 "  //x\n": HiddenRegionReplacer: space=' ';newLine=0
```

The issue here is that my test `TabAndSpacesSupportingWhiteSpaceReplacer` is not smart enough. It's not really meant to provide the correct answer to this test (it was to illustrate the issue in ML comments). It naively always adds the `space` which was set to one by the `oneSpace`. This is normally prevented by using a separate boolean for space indentation (instead of reusing `space`) and then making sure the spaceIndentation is decreased at the end of the comment such as done in step 1 in the HiddenRegionReplacer.

Reply:
Never mind, I've just made it a little less naive